File: lib/telegram_direct.js
// Direct Telegram send — bypasses n8n entirely (ADR-085's JS counterpart
// to channels/telegram_direct.py). Every existing Telegram notification
// routed through n8n (ADR-073) depends on n8n both running and its
// 04_Telegram_Notify workflow staying toggled Active — a real, repeatedly-
// confirmed one-click platform requirement (ADR-072) that a stale browser
// tab has already silently reverted once (ADR-073's own postmortem). This
// module sends the same real, critical events directly via Telegram's Bot
// API instead, so they land even when n8n is down or its workflow has
// drifted inactive.
//
// Deliberately fail-safe, same contract as lib/n8n_notify.js's
// notifyN8nProductionEvent(): missing config or a network failure is
// reported, never thrown — a real event (a sale happened, a product is
// ready, the factory hit a real error) must never be lost or crash its
// caller just because a Telegram send failed.

const DEFAULT_TIMEOUT_MS = 5000;

async function sendTelegramMessage(text, {
  token = process.env.TELEGRAM_BOT_TOKEN,
  chatId = process.env.OPENCLAW_TELEGRAM_CHAT_ID,
  timeoutMs = DEFAULT_TIMEOUT_MS,
  log = () => {},
} = {}) {
  if (!token || !chatId) {
    log({ event: 'skipped', reason: 'TELEGRAM_BOT_TOKEN or OPENCLAW_TELEGRAM_CHAT_ID not configured' });
    return { sent: false, message_id: null, error: 'TELEGRAM_BOT_TOKEN or OPENCLAW_TELEGRAM_CHAT_ID not configured' };
  }

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ chat_id: chatId, text }),
      signal: controller.signal,
    });
    const body = await res.json().catch(() => ({}));
    if (!res.ok || !body.ok) {
      const error = (body && body.description) || `HTTP ${res.status}`;
      log({ event: 'failed', error });
      return { sent: false, message_id: null, error };
    }
    log({ event: 'succeeded', message_id: body.result && body.result.message_id });
    return { sent: true, message_id: body.result && body.result.message_id, error: null };
  } catch (err) {
    log({ event: 'error', error: err.message });
    return { sent: false, message_id: null, error: err.message };
  } finally {
    clearTimeout(timer);
  }
}

// ADR-073's standing principle applied to the 3 events it named but left
// unwired: clear Arabic, key numbers, no English-then-translated text.

function buildProductReadyMessage(dossier) {
  const price = dossier.pricing_strategy && dossier.pricing_strategy.recommended_price;
  const lines = [
    '📄 منتج جاهز للمراجعة',
    '',
    `النيتش: ${dossier.niche}`,
  ];
  if (price != null) lines.push(`السعر الموصى به: $${price}`);
  lines.push(`رقم الإنتاج: ${dossier.production_id}`);
  return lines.join('\n');
}

function buildSaleMadeMessage(sale) {
  const lines = [
    '💰 بيع جديد!',
    '',
    `المنصة: ${sale.platform}`,
  ];
  if (sale.amount != null) lines.push(`المبلغ: $${sale.amount}`);
  if (sale.title) lines.push(`المنتج: ${sale.title}`);
  return lines.join('\n');
}

function buildCriticalErrorMessage(reasons) {
  const lines = [
    '🚨 عطل حرج',
    '',
    reasons[0],
  ];
  if (reasons.length > 1) lines.push(`(+${reasons.length - 1} سبب إضافي — راجع NEEDS_ATTENTION.md)`);
  return lines.join('\n');
}

module.exports = {
  sendTelegramMessage,
  buildProductReadyMessage,
  buildSaleMadeMessage,
  buildCriticalErrorMessage,
  DEFAULT_TIMEOUT_MS,
};

File: channels/telegram_direct.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Galaxy Forge — direct Telegram send (bypasses n8n entirely).

Every existing Telegram notification in this factory (ADR-073's Golden
Hunter Bridge message) goes browser->n8n->Telegram, and depends on n8n
being both running and its `04_Telegram_Notify` workflow toggled Active
(ADR-072: a real, confirmed one-click platform requirement n8n itself
enforces, not something this factory's code controls). ADR-072's and
ADR-074's addenda already sent two real, critical one-off messages by
calling Telegram's Bot API directly instead — this module is that same
proven pattern turned into a reusable function, rather than a third
one-off hand-written call.

Reads TELEGRAM_BOT_TOKEN / OPENCLAW_TELEGRAM_CHAT_ID from .env, same
lookup order as paddle_publisher.load_api_key(). Never raises to the
caller — a failed send is reported, not thrown, matching
lib/n8n_notify.js's notifyN8nProductionEvent() fail-safe philosophy: a
real, already-true event (money is real, a checkout link exists) must
never be lost or crash a caller just because Telegram is unreachable.
"""

import os
from pathlib import Path

import requests

FACTORY_DIR = Path(__file__).resolve().parent.parent
DEFAULT_ENV_PATH = FACTORY_DIR / ".env"
TELEGRAM_API_BASE = "https://api.telegram.org"


def _load_env_value(name, env_path=None):
    value = os.environ.get(name)
    if value:
        return value
    env_path = Path(env_path) if env_path else DEFAULT_ENV_PATH
    if env_path.exists():
        with open(env_path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line.startswith(name):
                    parsed = line.split("=", 1)[1].strip()
                    if parsed:
                        return parsed
    return None


def send_telegram_message(text, env_path=None, timeout=15):
    """Sends `text` to the founder's real Telegram chat. Returns
    {"sent": bool, "message_id": int|None, "error": str|None} — never
    raises. Missing config is reported as an error, not a crash, same
    honesty-over-silence discipline as every other channel adapter."""
    token = _load_env_value("TELEGRAM_BOT_TOKEN", env_path)
    chat_id = _load_env_value("OPENCLAW_TELEGRAM_CHAT_ID", env_path)
    if not token or not chat_id:
        return {"sent": False, "message_id": None, "error": "TELEGRAM_BOT_TOKEN or OPENCLAW_TELEGRAM_CHAT_ID not configured"}

    try:
        r = requests.post(
            f"{TELEGRAM_API_BASE}/bot{token}/sendMessage",
            json={"chat_id": chat_id, "text": text},
            timeout=timeout,
        )
        body = r.json() if r.content else {}
    except requests.RequestException as e:
        return {"sent": False, "message_id": None, "error": f"Telegram request failed: {e}"}

    if not r.ok or not body.get("ok"):
        return {"sent": False, "message_id": None, "error": body.get("description") or f"HTTP {r.status_code}"}

    return {"sent": True, "message_id": (body.get("result") or {}).get("message_id"), "error": None}
